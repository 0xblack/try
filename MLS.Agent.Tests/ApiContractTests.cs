using System;
using System.Threading.Tasks;
using Assent;
using WorkspaceServer.Models;
using WorkspaceServer.Models.Execution;
using WorkspaceServer.Tests;
using Xunit;
using Xunit.Abstractions;

namespace MLS.Agent.Tests
{
    public class ApiContractTests : ApiViaHttpTestsBase
    {
        private readonly Configuration configuration;

        public ApiContractTests(ITestOutputHelper output) : base(output)
        {
            configuration = new Configuration()
                            .UsingExtension("json");

#if !DEBUG
            configuration = configuration.SetInteractive(false);
#endif
        }

        [Fact]
        public async Task The_Run_contract_for_compiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode();

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id,
                position: viewport.Position);

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Run_contract_for_noncompiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode("doesn't compile");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id,
                position: viewport.Position);

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        private static Workspace.Buffer EntrypointCode(string mainContent = @"Console.WriteLine(Sample.Method());$$")
        {
            var input = $@"
using System;
using System.Linq;

namespace Example
{{
    public class Program
    {{
        public static void Main()
        {{
            {mainContent}
        }}       
    }}
}}";

            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "EntrypointCode.cs",
                output,
                position ?? 0);
        }

        private static Workspace.Buffer ViewportCode(string methodContent = @"return ""Hello world!"";$$ ")
        {
            var input = $@"
using System.Collections.Generic;
using System;

namespace Example
{{
    public static class Sample
    {{
        public static object Method()
        {{
#region viewport
            {methodContent}
#endregion
        }}
    }}
}}";

            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "ViewportCode.cs",
                output,
                position ?? 0);
        }
    }

}
