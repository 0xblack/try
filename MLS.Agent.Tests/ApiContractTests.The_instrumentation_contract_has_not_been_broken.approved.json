{
  "diagnostics": [
    {
      "start": 1,
      "end": 14,
      "message": "Unnecessary using directive.",
      "severity": 0,
      "id": "CS8019"
    },
    {
      "start": 15,
      "end": 33,
      "message": "Unnecessary using directive.",
      "severity": 0,
      "id": "CS8019"
    }
  ],
  "succeeded": true,
  "output": [],
  "exception": null,
  "instrumentation": [
    {
      "stackTrace": "Program Started",
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "filePosition": {
        "line": 10,
        "character": 12,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "filePosition": {
        "line": 10,
        "character": 23,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [
        {
          "name": "a",
          "value": "1",
          "declaredAt": {
            "start": 146,
            "end": 147
          }
        }
      ],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "filePosition": {
        "line": 10,
        "character": 34,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [
        {
          "name": "b",
          "value": "2",
          "declaredAt": {
            "start": 157,
            "end": 158
          }
        },
        {
          "name": "a",
          "value": "1",
          "declaredAt": {
            "start": 146,
            "end": 147
          }
        }
      ],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "filePosition": {
        "line": 10,
        "character": 41,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [
        {
          "name": "b",
          "value": "2",
          "declaredAt": {
            "start": 157,
            "end": 158
          }
        },
        {
          "name": "a",
          "value": "3",
          "declaredAt": {
            "start": 146,
            "end": 147
          }
        }
      ],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "stackTrace": "Program Terminated",
      "output": {
        "start": 0,
        "end": 0
      }
    }
  ],
  "variableLocations": [
    {
      "name": "b",
      "locations": [
        {
          "startLine": 10,
          "startColumn": 41,
          "endLine": 10,
          "endColumn": 42
        },
        {
          "startLine": 10,
          "startColumn": 27,
          "endLine": 10,
          "endColumn": 32
        }
      ],
      "declaredAt": {
        "start": 157,
        "end": 162
      }
    },
    {
      "name": "a",
      "locations": [
        {
          "startLine": 10,
          "startColumn": 34,
          "endLine": 10,
          "endColumn": 35
        },
        {
          "startLine": 10,
          "startColumn": 45,
          "endLine": 10,
          "endColumn": 46
        },
        {
          "startLine": 10,
          "startColumn": 16,
          "endLine": 10,
          "endColumn": 21
        }
      ],
      "declaredAt": {
        "start": 146,
        "end": 151
      }
    }
  ]
}